Tests/WorkspaceTests/WorkspaceTests.swift

```diff
@@ -4663,6 +4663,524 @@ final class WorkspaceTests: XCTestCase {
+    func testDuplicateTransitiveIdentityWithoutNames() throws {
+        let sandbox = AbsolutePath("/tmp/ws/")
+        let fs = InMemoryFileSystem()
+
+        let workspace = try MockWorkspace(
+            sandbox: sandbox,
+            fs: fs,
+            roots: [
+                MockPackage(
+                    name: "Root",
+                    targets: [
+                        MockTarget(name: "RootTarget", dependencies: [
+                            .product(name: "FooUtilityProduct", package: "FooUtilityPackage"),
+                            .product(name: "BarProduct", package: "BarPackage")
+                        ]),
+                    ],
+                    products: [],
+                    dependencies: [
+                        MockDependency(name: nil, path: "foo/utility", requirement: .upToNextMajor(from: "1.0.0")),
+                        MockDependency(name: nil, path: "bar", requirement: .upToNextMajor(from: "1.0.0")),
+                    ],
+                    toolsVersion: .v5
+                ),
+            ],
+            packages: [
+                MockPackage(
+                    name: "FooUtilityPackage",
+                    path: "foo/utility",
+                    targets: [
+                        MockTarget(name: "FooUtilityTarget"),
+                    ],
+                    products: [
+                        MockProduct(name: "FooUtilityProduct", targets: ["FooUtilityTarget"]),
+                    ],
+                    versions: ["1.0.0"]
+                ),
+                MockPackage(
+                    name: "BarPackage",
+                    path: "bar",
+                    targets: [
+                        MockTarget(name: "BarTarget", dependencies: [
+                            .product(name: "OtherUtilityProduct", package: "OtherUtilityPackage"),
+                        ]),
+                    ],
+                    products: [
+                        MockProduct(name: "BarProduct", targets: ["BarTarget"]),
+                    ],
+                    dependencies: [
+                        MockDependency(name: nil, path: "other-foo/utility", requirement: .upToNextMajor(from: "1.0.0")),
+                    ],
+                    versions: ["1.0.0"]
+                ),
+                // this package never gets loaded since its identity is the same as "FooPackage"
+                MockPackage(
+                    name: "OtherUtilityPackage",
+                    path: "other/utility",
+                    targets: [
+                        MockTarget(name: "OtherUtilityTarget"),
+                    ],
+                    products: [
+                        MockProduct(name: "OtherUtilityProduct", targets: ["OtherUtilityTarget"]),
+                    ],
+                    versions: ["1.0.0"]
+                ),
+            ]
+        )
+
+        // FIXME: rdar://72940946
+        // we need to improve this situation or diagnostics when working on identity
```

Comment:
👀 this is a fun one we dont deal with well right now given URL based identity collisions